### PR TITLE
Add Debug builds to CI

### PR DESCRIPTION
**Pull request by michdolan**
This PR resolves some issues with Debug builds failing when using OCIO_INSTALL_EXT_PACKAGES, and adds a Debug build to each platform in Azure Pipelines.

### CODE REVIEW

**Comment by michdolan** [comment]
@hodoulp , @BernardLefebvre , when you have cycles can you give this a try in your Windows build env?

This is building Debug properly in Azure for Linux and macOS, and successfully linking all of the dependencies pulled with OCIO_INSTALL_EXT_PACKAGES on Windows. The Azure Windows build fails in Debug still because the agent doesn't have python27_d.lib, which Python is hardcoded to use for Debug linking (see: https://answers.opencv.org/question/68148/lnk1104-cannot-open-file-python34_dlib/). I opted to not build Python on Azure to test this, but assuming you have a proper debug Python lib, it should be a non-issue in practice.

I did have to update ilmbase, expat, and yaml-cpp to use their respective Debug suffixes, and adjust the Half install usage. I think the errors you were seeing were related to these issues.

**Comment by michdolan** [comment]
Also to note: I swapped the PythonInterp/PythonLibs find_package calls because, according to CMake docs, order matters.

**Comment by hodoulp** [comment]
@michdolan Knowing that the debug build is quite often broken any time major works are done on the CMake files, I would recommend to keep one debug build per platform (i.e. Windows & Linux). 

Note: The Windows Debug CI build could not compile the Python bindings and the documentation to avoid the Python debug install 'issue'. Or a later change could be done to use the Python release libraries when compiling in debug as I do not see any use case where someone would like to debug the Python libraries.

File: share/cmake/modules/FindIlmBase.cmake [hunk 1]
```diff
@@ -56,15 +56,26 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
...
+        )
+        if(CMAKE_BUILD_TYPE STREQUAL Debug)
+            # WIN32 <= 2.3.0, All platforms > 2.3.0
+            list(INSERT _ILMBASE_STATIC 0
+                "${CMAKE_STATIC_LIBRARY_PREFIX}Half-${_ILMBASE_LIB_VER}_s_d${CMAKE_STATIC_LIBRARY_SUFFIX}")
+        endif()
     endif()
 
     # Find library
+    if(CMAKE_BUILD_TYPE STREQUAL Debug)
```
Comment (hodoulp) [inline share/cmake/modules/FindIlmBase.cmake]: Could you explain a little bit these lines? Is that related to some `Azure` limitations?

File: share/cmake/modules/FindYamlCpp.cmake [hunk 2]
```diff
@@ -114,7 +125,10 @@ if(NOT YAMLCPP_FOUND)
     set(YAMLCPP_VERSION ${YamlCpp_FIND_VERSION})
     set(YAMLCPP_INCLUDE_DIR "${_EXT_DIST_ROOT}/include")
     if(WIN32)
-        set(YAMLCPP_LIBRARY "${_EXT_DIST_ROOT}/lib/libyaml-cppmd.lib")
+        if(CMAKE_BUILD_TYPE STREQUAL Debug)
+            set(_YAMLCPP_LIB_SUFFIX "d")
+        endif()
+        set(YAMLCPP_LIBRARY "${_EXT_DIST_ROOT}/lib/libyaml-cppmd${_YAMLCPP_LIB_SUFFIX}.lib")
```
Comment (hodoulp) [inline share/cmake/modules/FindYamlCpp.cmake]: You could use `${CMAKE_STATIC_LIBRARY_SUFFIX}` instead of `.lib`or `.a`.

File: share/cmake/modules/FindExpat.cmake [hunk 3]
```diff
@@ -59,12 +59,20 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
     # Attempt to find static library first if this is set
     if(EXPAT_STATIC_LIBRARY)
         set(_EXPAT_STATIC "${CMAKE_STATIC_LIBRARY_PREFIX}expat${CMAKE_STATIC_LIBRARY_SUFFIX}")
+        if(WIN32 AND CMAKE_BUILD_TYPE STREQUAL Debug)
```
[the remark this answers is not in the document]
Comment (michdolan) [reply share/cmake/modules/FindExpat.cmake]: In the CMake systems we depend on, the different suffix only applies when doing a full Debug build. This is why I used Debug as the condition check instead of Release, so that the Release behavior would apply to the other non-Debug types. I will double check that though with each project.

File: share/cmake/modules/FindIlmBase.cmake (hunk 1 above)
Comment (michdolan) [reply share/cmake/modules/FindIlmBase.cmake]: I'll improve this comment. As an explanation, OpenEXR 2.3.0 and below only add the _d suffix when building for Windows. I observed however that CMake in the current master is being overhauled, and this has been changed so that _d is appended to all Debug builds. I'll double check this.

**Comment by michdolan** [comment]
It is possible to hack the agent's Python config header to make it use the Release Python lib instead (it's hard coded in Python install itself, regardless of supplying the correct path to CMake), but I was apprehensive about relying on that. I'll disable Python in that build for now.

File: share/cmake/modules/FindExpat.cmake [hunk 4]
```diff
@@ -59,12 +59,20 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
     # Attempt to find static library first if this is set
     if(EXPAT_STATIC_LIBRARY)
         set(_EXPAT_STATIC "${CMAKE_STATIC_LIBRARY_PREFIX}expat${CMAKE_STATIC_LIBRARY_SUFFIX}")
```
[the remark this answers is not in the document]
Comment (michdolan) [reply share/cmake/modules/FindExpat.cmake]: I restructured the lib name list construction. Hopefully it's more clear now. Since this part of the code is creating a prioritised list of lib names to search for that *might* already be installed on the system, it needs to handle more than one case. i.e. not failing if the user has a Release version of a dependency, even though they are building OCIO in Debug. The code prioritises Debug libs in this case, but it *can* fallback to other variations.

File: share/cmake/modules/FindIlmBase.cmake [hunk 5]
```diff
@@ -56,15 +56,26 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
 
     # Attempt to find static library first if this is set
     if(ILMBASE_STATIC_LIBRARY)
```
[the remark this answers is not in the document]
Comment (michdolan) [reply share/cmake/modules/FindIlmBase.cmake]: Reorganised based on comments above

File: share/cmake/modules/FindYamlCpp.cmake [hunk 6]
```diff
@@ -58,13 +58,24 @@ if(NOT OCIO_INSTALL_EXT_PACKAGES STREQUAL ALL)
 
     # Attempt to find static library first if this is set
```
[the remark this answers is not in the document]
Comment (michdolan) [reply share/cmake/modules/FindYamlCpp.cmake]: Reorganised based on comments above

File: share/cmake/modules/FindIlmBase.cmake (hunk 1 above)
Comment (michdolan) [reply share/cmake/modules/FindIlmBase.cmake]: I expounded on this in the install portion of the Find module, and removed earlier instances of the comment where it was less necessary (just a possible known lib name).

File: share/cmake/modules/FindYamlCpp.cmake (hunk 2 above)
Comment (michdolan) [reply share/cmake/modules/FindYamlCpp.cmake]: Fixed.

**Comment by michdolan** [comment]
I added a Debug build to each platform in Azure, and disabled Python/docs in the Windows Debug build with a TODO.

File: share/cmake/modules/FindExpat.cmake (hunk 3 above)
[the remark this answers is not in the document]
Comment (hodoulp) [reply share/cmake/modules/FindExpat.cmake]: 1. ok.
2. What about `CMAKE_BUILD_TYPE` equal to `debug`(i.e. all in lower case)? 
The `STREQUAL` is case-sensitive.

[the remark this answers is not in the document]
Comment (michdolan) [reply share/cmake/modules/FindExpat.cmake]: Fixed

**Comment by michdolan** [comment]
MATCHES would require comparison with "[Dd][Ee][Bb][Uu][Gg]" to be case insensitive. I'll define it in the root cmake.

**Comment by michdolan** [comment]
I could add a macro for case insensitive matching in general, which could be reused throughout.
